conda/common/logic.py

```diff
@@ -29,26 +29,195 @@
+    def invoke(self, iter_sol):
+        # sat_solution = pycosat.solve(clauses.as_list(), vars=m, prop_limit=limit)
+        try:
+            sat_solution = next(iter_sol)
+        except StopIteration:
+            sat_solution = "UNSAT"
+        del iter_sol
+        return sat_solution
+
+    @time_recorder(module_name=__name__)
+    def process_solution(self, sat_solution):
+        if sat_solution in ("UNSAT", "UNKNOWN"):
+            return None
+        return sat_solution
+
+
+class CryptoMiniSatSolver(SatSolver):
+    @time_recorder(module_name=__name__)
+    def setup(self, clauses, m, threads=1):
+        from pycryptosat import Solver
+        solver = Solver(threads=threads)
+        solver.add_clauses(clauses.as_list())
+        return solver
+
+    @time_recorder(module_name=__name__)
+    def invoke(self, solver):
+        sat, sat_solution = solver.solve()
+        if not sat:
+            sat_solution = None
+        return sat_solution
+
+    @time_recorder(module_name=__name__)
+    def process_solution(self, solution):
+        if not solution:
+            return None
+        # first element of solution is always None
+        solution = [i for i, b in enumerate(solution) if b]
+        return solution
+
+
+class PySatSolver(SatSolver):
+    @time_recorder(module_name=__name__)
+    def setup(self, clauses, m, **kwargs):
+        from pysat import solvers
+        solver = solvers.Glucose4()
+        # FIXME upstream: pysat.solvers require a clause to be a list...
+        clauses = list(map(list, clauses.as_list()))
+        solver.append_formula(clauses)
+        return solver
+
+    @time_recorder(module_name=__name__)
+    def invoke(self, solver):
+        if not solver.solve():
+            sat_solution = None
+        else:
+            sat_solution = solver.get_model()
+        solver.delete()
+        return sat_solution
+
+    @time_recorder(module_name=__name__)
+    def process_solution(self, sat_solution):
+        if sat_solution is None:
+            solution = None
+        else:
+            solution = sat_solution
+        return solution
+
+
 # Code that uses special cases (generates no clauses) is in ADTs/FEnv.h in
 # minisatp. Code that generates clauses is in Hardware_clausify.cc (and are
 # also described in the paper, "Translating Pseudo-Boolean Constraints into
 # SAT," Eén and Sörensson).
 class Clauses(object):
     def __init__(self, m=0):
-        self.clauses = []
         self.names = {}
         self.indices = {}
         self.unsat = False
         self.m = m
+        self._clauses = ClauseList()
+        # bind some methods of _clauses to reduce call overhead
+        self.add_clause = self._clauses.append
+        self.add_clauses = self._clauses.extend
+        self.get_clause_count = self._clauses.get_clause_count
+        self.save_state = self._clauses.save_state
+        self.restore_state = self._clauses.restore_state
+        self.as_list = self._clauses.as_list
```

Comment:
If we wanted to make `self._clauses` re-assignable, we can just implement setting these in `__setattr__` or something.